### ISSUE TITLE
continous instead of binary trading decision?

### ISSUE THREAD

**Issue by saets** [askmike/gekko#497]
Is there a way to use continous trading decisions instead of just having the binary choice of "long" or "short"?

A trading system where you can only choose between going long or short, but not scale into or out of positions is not usefull, unless you are only trading such small amounts, that it actually makes no difference.

more precisely, what you want to do is to balance the risk exposure you are taking with a position at a given time. you do this by calculating the risk of holding a position (e.g.: via recent price volatility) and adjust your position size according to your portfolio holdings of cash and assets (or several assets).

sounds complicated, but is relatively easy if you know what your portfolio looks like (portfolio object?) and ofcourse you need to be able to place limit orders.

long story short: switching between just  BUY, HOLD, and SELL is good enough for casual small scale trading, but if you are serious about it it is a dead end.

Can this be done and if, how?

p.s.: I bow to askmike for his work on this project so far. It is great.

**Comment by askmike** [comment]
Hey Seats,

You are definitely right. Gekko started out as a very simple "script" to watch Mt. Gox BTC/USD and advice based on MACD over hourly candles. I guess it all grew from there.

> more precisely, what you want to do is to balance the risk exposure you are taking with a position at a given time. you do this by calculating the risk of holding a position (e.g.: via recent price volatility) and adjust your position size according to your portfolio holdings of cash and assets (or several assets).

While you are 100% correct I do think it is important to note two things:
- Managing risk is extremely important, however the ways in which you can do this at most cryptocurrency exchanges are very limited. Since most cryptocurrencies are heavily correlated (to bitcoin) hedging risk or diversifying a portfolio is not really possible in most cases (except for some altcoin exceptions).
- One of the biggest risks dealing with cryptocurrencies is the keeping your funds in an exchange in the first place (I've been burned on Mt. Gox, Cryptsy and Bitfinex).

---

> you do this by calculating the risk of holding a position (e.g.: via recent price volatility) and adjust your position size according to your portfolio holdings of cash and assets

When your strategy calls `this.advice` that advice is actually wrapped in an object, there is a property called portfolio (see [here](https://github.com/askmike/gekko/blob/develop/plugins/tradingAdvisor/baseTradingMethod.js#L265)) that indicates how much of portfolio should be moved into the direction of the advice.

However this is ignored by the Trader (code that uses this advice to place orders) and the paper trader (code that simulates orders).

Would something like this satisfy what you want?

```
this advice('long', 0.01) // one percent of portfolio should move into a long position
```

**Comment by saets** [comment]
Hey askmike,

thanks for your fast reply!

I think its awsome, that this project has grown over this long period of time and still be relevant and alive.
If I can use it with my still very limited coding skills, I would love to be part of it and maybe also make some meaningful suggestions. 

Concerning risk management with portfolio diversification, I totally agree with you. It is not worth the effort most of the time. 
What I meant was managing exposure within one trading pair (on one exchange) according to the recent (lets say 10 day) volatility of that pair.
Because (nearly) any indicator signal will have a different success rate in different market conditions (e.g. uptrend, sideways, downtrend each with much or little volatility - giving you six different market "weathers").
So for instance if I know from (good and out of sample backtesting...) that one part of my trading system will give me good signals when market goes sideways with high volatility, but gets destroyed when as soon as the volatility drops, I want to give less weight to the signals if and when that happens. 

Long story short: I have a system, that can tell me how much of my portfolio should be in a long (or short) position. 
I then compare that to my actual portfolio position and trade the difference.

What I would need is having access to my portfolio (how much btc, how much fiat I hold) and then place a limit order for the difference to what my strategy tells me I should hold (ideally with a timeout).

e.g.: holding 5 BTC 2500 USD at an example price of 500 USD. I am long with 50% of my equity value. Now my strategy thinks the price will rise further and gives me a  a target position of 70% long.
I now calculate how much to buy at current price to be in a 70% long position. 
And place a limit order. 

(How) can I do that?

best,

saets.

p.s.:
I got my last btc out of gox 10 days before it went to... only to later get burned at hitbtc, and ofcourse bitfinex.

**Comment by askmike** [comment]
Sorry for the late response.

> So for instance if I know from (good and out of sample backtesting...) that one part of my trading system will give me good signals when market goes sideways with high volatility, but gets destroyed when as soon as the volatility drops, I want to give less weight to the signals if and when that happens.

Can you not simply measure volatility in your method and if it drops not advice?

> What I would need is having access to my portfolio (how much btc, how much fiat I hold) and then place a limit order for the difference to what my strategy tells me I should hold (ideally with a timeout).

Since this issue opened I have been working a lot on Gekko and introduced a lot new functionality (including an webinterface). Unfortunately this request is not new functionality.

The hard thing about this is that communication between different parts is tightly controlled: The only communication between the strat and the trader (or paperTrader) right now is sending advice from the strat to the trader. This would require a way for the strat to query the trader for information about current portfolioHoldings.

I think this would not only make Gekko's code more complex, but also the strategy's API surface.

----

To come back on the question: 

> Would something like this satisfy what you want?

>     this advice('long', 0.01) // one percent of portfolio should move into a long position

As soon as you or anyone else replies I'll reopen this.

**Comment by osarrouy** [comment]
Hi Mike,

First of all thanks for all your fantastic job on Gekko.

I answer there to reopen the thread. 

I would personally be really pleased to get something like the feature you propose :

`this advice('long', 0.01) // one percent of portfolio should move into a long position`

Another question. Would it be possible - and if yes : how ? :) - to access the current portfolio from the inside of a strategy.

I think i would be great not only to define how much you want to place on such or such order, but also to define things such as stop loss from the inside of a given strategy.

Thanks again for your work.

**Comment by tim-mills** [comment]
I'd be really keen to see this enhancement - there's no current way to do this now is there?
